backend/src/create/dog_leg.rs

```diff
@@ -0,0 +1,157 @@
+use geo::{Euclidean, InterpolatableLine, Length};
+use utils::osm2graph::{EdgeID, Graph, Intersection, IntersectionID};
+
+use crate::geo_helpers::bearing_from_endpoint;
+
+/// Dog-leg intersections are "nearly" 4-way intersections, where a short edge separates two 3-way
+/// intersections. Collapse them into a regular 4-way intersection so diagonal filters can be used.
+pub fn collapse_dog_legs(graph: &mut Graph) {
+    let mut dog_legs = Vec::new();
+    for e in graph.edges.keys() {
+        if let Some((side1, side2)) = detect_dog_leg(graph, *e) {
+            dog_legs.push((*e, side1, side2));
+        }
+    }
+    info!("Collapsing {} dog-legs", dog_legs.len());
+
+    for (collapse_e, side1, side2) in dog_legs {
+        fix_dog_leg(graph, collapse_e, side1, side2);
+    }
+
+    // Remove deleted edges from node_to_edge, used later for applying existing modal filters.
+    graph
+        .node_to_edge
+        .retain(|_, e| graph.edges.contains_key(e));
+}
+
+/// If this edge is the short edge causing a dog-leg intersection, then return the two
+/// perpendicular "side roads".
+fn detect_dog_leg(graph: &Graph, e: EdgeID) -> Option<(EdgeID, EdgeID)> {
+    let edge = &graph.edges[&e];
+    if Euclidean.length(&edge.linestring) > 5.0 {
+        return None;
+    }
+    if !edge.osm_tags.has("name") {
+        return None;
+    }
+    let src_i = &graph.intersections[&edge.src];
+    let dst_i = &graph.intersections[&edge.dst];
+    let mut src_edges = src_i.edges.clone();
+    let mut dst_edges = dst_i.edges.clone();
```

Comment:
It might not be a big deal, but it think we could avoid these clones if this method is expensive.